internal/rbd/controllerserver.go

```diff
@@ -622,49 +731,81 @@ func (cs *ControllerServer) validateSnapshotReq(ctx context.Context, req *csi.Cr
+	// generate cloned volume details from snapshot
+	cloneRbd := generateVolFromSnap(rbdSnap)
+	defer cloneRbd.Destroy()
+	// add image feature for cloneRbd
+	f := []string{librbd.FeatureNameLayering, librbd.FeatureNameDeepFlatten}
+	cloneRbd.imageFeatureSet = librbd.FeatureSetFromNames(f)
+	ready := false
+
+	err := cloneRbd.Connect(cr)
+	if err != nil {
+		return ready, cloneRbd, err
+	}
+
+	err = createRBDClone(ctx, parentVol, cloneRbd, rbdSnap, cr)
 	if err != nil {
 		klog.Errorf(util.Log(ctx, "failed to create snapshot: %v"), err)
-		return status.Error(codes.Internal, err.Error())
+		return ready, cloneRbd, status.Error(codes.Internal, err.Error())
 	}
+
 	defer func() {
 		if err != nil {
-			errDefer := deleteSnapshot(ctx, rbdSnap, cr)
-			if errDefer != nil {
-				klog.Errorf(util.Log(ctx, "failed to delete snapshot: %v"), errDefer)
-				err = fmt.Errorf("snapshot created but failed to delete snapshot due to"+
-					" other failures: %v", err)
+			if _, ok := err.(ErrFlattenInProgress); !ok {
+				// cleanup clone and snapshot
+				errCleanUp := cleanUpSnapshot(ctx, cloneRbd, rbdSnap, cloneRbd, cr)
+				if errCleanUp != nil {
+					klog.Errorf(util.Log(ctx, "failed to cleanup snapshot and clone: %v"), errCleanUp)
+				}
 			}
-			err = status.Error(codes.Internal, err.Error())
 		}
 	}()
-	err = protectSnapshot(ctx, rbdSnap, cr)
+
+	err = createSnapshot(ctx, cloneRbd, rbdSnap)
 	if err != nil {
-		klog.Errorf(util.Log(ctx, "failed to protect snapshot: %v"), err)
-		return status.Error(codes.Internal, err.Error())
+		// update rbd image name for logging
+		rbdSnap.RbdImageName = cloneRbd.RbdImageName
+		klog.Errorf(util.Log(ctx, "failed to create snapshot %s: %v"), rbdSnap, err)
+		return ready, cloneRbd, err
 	}
-	defer func() {
-		if err != nil {
-			errDefer := unprotectSnapshot(ctx, rbdSnap, cr)
-			if errDefer != nil {
-				klog.Errorf(util.Log(ctx, "failed to unprotect snapshot: %v"), errDefer)
-				err = fmt.Errorf("snapshot created but failed to unprotect snapshot due to"+
-					" other failures: %v", err)
-			}
-			err = status.Error(codes.Internal, err.Error())
-		}
-	}()
 
-	err = getSnapshotMetadata(ctx, rbdSnap, cr)
+	err = cloneRbd.getImageID()
 	if err != nil {
-		klog.Errorf(util.Log(ctx, "failed to fetch snapshot metadata: %v"), err)
-		return status.Error(codes.Internal, err.Error())
+		klog.Errorf(util.Log(ctx, "failed to get image id: %v"), err)
+		return ready, cloneRbd, err
 	}
 
-	return nil
+	err = cloneRbd.updateVolWithImageInfo(ctx, cr)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to update image details for %s: %v"), cloneRbd, err)
+		return ready, cloneRbd, err
+	}
+	var j = &journal.Connection{}
+	// save image ID
+	j, err = snapJournal.Connect(rbdSnap.Monitors, cr)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to connect to cluster: %v"), err)
+		return ready, cloneRbd, err
+	}
+	defer j.Destroy()
+
+	err = j.ReserveImageID(ctx, rbdSnap.JournalPool, rbdSnap.ReservedID, cloneRbd.ImageID, cr)
```

Comment:
If we fail just prior to this call (as in provisioner was restarted, say), we would fail to save the image ID. The snapshot although is complete, so on a `CreateSnapshot` retry we would succeed in `checkSnapCloneExists` and not save the image ID for `DeleteSnapshot` call.

Either `checkSnapCloneExists` should ensure the image ID is saved, or undo the entire work, or the image ID should be saved sooner in the transaction to prevent such cases.